epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt

```diff
@@ -35,126 +35,135 @@ import java.util.concurrent.Executor
-   */
-  private var lastPosition: Int? = null
+    /**
+     * Backing list for built models. This is a full array list that has null items for not yet build models.
+     *
+     * All interactions with this should by synchronized, since it is accessed from several threads
+     * for model building, paged list updates, and cache clearing.
+     */
+    private val modelCache = arrayListOf<EpoxyModel<*>?>()
+    /**
+     * Tracks the last accessed position so that we can report it back to the paged list when models are built.
+     */
+    private var lastPosition: Int? = null
 
-  /**
-   * Observer for the PagedList changes that invalidates the model cache when data is updated.
-   */
-  private val updateCallback = object : ListUpdateCallback {
-    override fun onChanged(position: Int, count: Int, payload: Any?) {
-      (position until (position + count)).forEach {
-        modelCache[it] = null
-      }
-      rebuildCallback()
-    }
-
-    override fun onMoved(fromPosition: Int, toPosition: Int) {
-        val model = modelCache.removeAt(fromPosition)
-        modelCache.add(toPosition, model)
-        rebuildCallback()
-    }
-
-    override fun onInserted(position: Int, count: Int) {
-        (0 until count).forEach { _ ->
-          modelCache.add(position, null)
+    /**
+     * Observer for the PagedList changes that invalidates the model cache when data is updated.
+     */
+    private val updateCallback = object : ListUpdateCallback {
+        @Synchronized
+        override fun onChanged(position: Int, count: Int, payload: Any?) {
+            (position until (position + count)).forEach {
+                modelCache[it] = null
+            }
+            rebuildCallback()
         }
-        rebuildCallback()
-    }
 
-    override fun onRemoved(position: Int, count: Int) {
-        (0 until count).forEach { _ ->
-          modelCache.removeAt(position)
+        @Synchronized
+        override fun onMoved(fromPosition: Int, toPosition: Int) {
+            val model = modelCache.removeAt(fromPosition)
+            modelCache.add(toPosition, model)
+            rebuildCallback()
         }
-        rebuildCallback()
-    }
-  }
 
-  private val asyncDiffer = @SuppressLint("RestrictedApi")
-  object : AsyncPagedListDiffer<T>(
-      updateCallback,
-      AsyncDifferConfig.Builder<T>(
-          itemDiffCallback
-      ).also {builder ->
-        if (diffExecutor != null) {
-          builder.setBackgroundThreadExecutor(diffExecutor)
+        @Synchronized
+        override fun onInserted(position: Int, count: Int) {
+            (0 until count).forEach {
+                modelCache.add(position, null)
+            }
+            rebuildCallback()
         }
-        // we have to reply on this private API, otherwise, paged list might be changed when models are being built,
-        // potentially creating concurrent modification problems.
-        builder.setMainThreadExecutor {runnable : Runnable ->
```

Comment:
this should be doing the thread sync.

Reply:
@yigit right, I understand that part (apologies for the indentation change from auto-format, it makes this hard to read, but most of the lines are the same.)

But as far as I can tell, this only guarantees synchronization when diffing a new paged list.

the main thing I'm trying to address with the synchronization is changes to the PagedList that get notified to the cache via the paged list's `notifyExecutor` (since the `AsyncPagedListDiffer`  uses `PagedList#addWeakCallback` to register `mPagedListCallback`)

You've said before that having people set their PagedList notifyExecutor to the same thread as model building would fix this, but that seems very fragile to me, and easy for people to miss.

Is there anything wrong with forcibly synchronizing the cache access across threads so we don't have to worry about the notifyExecutor?